### PR TITLE
Add support for Initiating User Registration via prompt=create (#10701)

### LINKED ISSUE #10701: AuthenticationRequest add "create" prompt
### Description

It would be nice to include the prompt = create parameter to redirect the user directly to the sign-up page.

As described here:
https://openid.net/specs/openid-connect-prompt-create-1_0.html#rfc.references.2

### Discussion

_No response_

### Motivation

_No response_

### Details

_No response_

### PR DESCRIPTION
**Pull request by thomasdarimont**
This PR adds support for "Initiating User Registration via OpenID Connect 1.0".

This allows to use the standard `prompt`parameter to initiate user registrations. Previously, users had to use
the propriatery registrations endpoint (`$KC_ISSUER/protocol/openid-connect/registrations`) for this.

See: https://openid.net/specs/openid-connect-prompt-create-1_0.html#section-4.1

Fixes #10701

### CODE REVIEW

**Comment by ahus1** [comment]
@keycloak/core-clients - this looks like "core-clients" to me. I wonder what happens if the realm doesn't allow registration. 

Looking at the spec, it mentions in the "Discovery Metadata" section: 

>If an OpenID Provider supports this specification it MUST define this metadata element in the openid-configuration file. 

So I suppose this is missing.

This reminds me that KC should have a list of supported standards/RFCs in its docs.

**Comment by thomasdarimont** [comment]
Thanks for the review! Good catch! I added the same check as we have in `org.keycloak.protocol.oidc.endpoints.AuthorizationEndpoint#register`, which shows the error page:
```
We are sorry... 
Registration not allowed 
```
when attempting a registration via `$KC_ISSUER/protocol/openid-connect/registrations` with disabled registration in Realm/Login
```
http://localhost:8081/auth/realms/demo/protocol/openid-connect/registrations?client_id=keycloak-website&redirect_uri=https%3A%2F%2Fwww.keycloak.org%2Fapp%2F%23url%3Dhttp%3A%2F%2Flocalhost%3A8081%2Fauth%26realm%3Ddemo%26client%3Dkeycloak-website&state=18ae93c0-1a77-4f03-a16e-d59c1c23f95c&response_mode=fragment&response_type=code&scope=openid&nonce=4f19f4ad-c784-4ae4-9933-1180d7f4d797&code_challenge=VAS5otlT-X3Z0bfc9dICaMde_I9brWWUp_9GuC3vX68&code_challenge_method=S256&prompt=create
````

The prompt=create behaves the same way now.

**Comment by thomasdarimont** [comment]
I also added the missing `prompt_values_supported` field to the OIDCWellKnown configuration representation.

File: services/src/main/java/org/keycloak/protocol/oidc/OIDCWellKnownProvider.java [hunk 1]
```diff
@@ -166,6 +168,7 @@ public Object getConfig() {
         config.setResponseModesSupported(DEFAULT_RESPONSE_MODES_SUPPORTED);
         config.setGrantTypesSupported(DEFAULT_GRANT_TYPES_SUPPORTED);
         config.setAcrValuesSupported(getAcrValuesSupported(realm));
+        config.setPromptValuesSupported(DEFAULT_PROMPT_VALUES_SUPPORTED);
```
Comment (ahus1) [inline services/src/main/java/org/keycloak/protocol/oidc/OIDCWellKnownProvider.java:171]: IMHO this should be realm specific. Due to that, the list of default prompt values should only contain those that are allowed for this realm. 

That is: Only if registration is allowed in this realm, the prompt `create` should be included in the list. 

Not sure if there could be conditions for the other values.

**Review by ahus1 — DISMISSED** [review]
Thank you for the well-known endpoint implementation. IMHO it needs some more refinement. See below.

File: services/src/main/java/org/keycloak/protocol/oidc/OIDCWellKnownProvider.java (hunk 1 above)
Comment (thomasdarimont) [reply services/src/main/java/org/keycloak/protocol/oidc/OIDCWellKnownProvider.java]: I made the generation of `prompt_values_supported` realm aware and only add the `create` scope when registrations are allowed for the realm.

Comment (ahus1) [reply services/src/main/java/org/keycloak/protocol/oidc/OIDCWellKnownProvider.java]: Thank you, this resolves it for me. I'll leave it for the core-client team to do a more in-depth review.

**Comment by thomasdarimont** [comment]
Note that the spec also performs a strict check on the prompt parameters in the authorization request.
In this PR I deliberately choose to not add this, as this might break existing users who use custom parameters for this.

We might decide to adopt a more strict validation later, with a new method like `org.keycloak.protocol.oidc.endpoints.AuthorizationEndpointChecker#checkPromptParam` e.g.:
```

    public void checkPromptParam() throws AuthorizationCheckException {
        boolean isOIDCRequest = TokenUtil.isOIDCRequest(request.getScope());
        if (!isOIDCRequest) {
            return;
        }

        String prompt = request.getPrompt();
        if (prompt == null) {
            return;
        }

        switch(prompt) {
            case OIDCLoginProtocol.PROMPT_VALUE_NONE:
            case OIDCLoginProtocol.PROMPT_VALUE_LOGIN:
            case OIDCLoginProtocol.PROMPT_VALUE_CONSENT:
            case OIDCLoginProtocol.PROMPT_VALUE_CREATE:
                break;
            default:
                throw new AuthorizationCheckException(Response.Status.BAD_REQUEST, OAuthErrorException.INVALID_REQUEST, "Supplied prompt parameter value is not supported");
        }
    }
```

**Review by mposolda — COMMENTED** [review]
@thomasdarimont Nice addition!

Is it possible to add also small update in this documentation section https://www.keycloak.org/docs/latest/server_admin/index.html#_registration-rc-client-flows ? I think we may mention that `prompt=create` is now preferred over the `/realms/your_realm/protocol/openid-connect/registrations` endpoint as that one is non-specs compliant. WDYT?
Maybe the `registrations` endpoint could be even marked as deprecated, so we have possibility to eventually remove it in the future releases.

**Comment by thomasdarimont** [comment]
@mposolda I added a small note to the documentation. I'm not sure if we should express a preference for `prompt=login` over `/registrations` as it might depend on what the OIDC client integration can do.

**Review by mposolda — APPROVED** [review]
@thomasdarimont Thanks!

**Pull request merged**